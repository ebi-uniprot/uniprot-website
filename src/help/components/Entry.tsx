import { RouteChildrenProps } from 'react-router-dom';

import SingleColumnLayout from '../../shared/components/layouts/SingleColumnLayout';

const HelpEntry = (props: RouteChildrenProps<{ accession: string }>) => (
  <SingleColumnLayout>
    Help entry: {props.match?.params.accession}
  </SingleColumnLayout>
);

export default HelpEntry;
