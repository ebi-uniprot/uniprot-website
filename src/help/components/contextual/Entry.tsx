import { RouteChildrenProps } from 'react-router-dom';

import HelpEntry from '../entry/Entry';

const Entry = (props: RouteChildrenProps<{ accession: string }>) => (
  <HelpEntry inPanel {...props} />
);

export default Entry;
