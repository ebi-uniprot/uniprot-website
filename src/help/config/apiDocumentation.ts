import { ApiDocsDefinition } from '../types/apiDocumentation';

export const apiDocsDefinitionToString = new Map([
  [ApiDocsDefinition.uniprotkb, 'UniProtKB'],
  [ApiDocsDefinition.uniref, 'UniRef'],
  [ApiDocsDefinition.uniparc, 'UniParc'],
  [ApiDocsDefinition.proteome, 'Proteomes'],
  [ApiDocsDefinition.support_data, 'Supporting Data'],
  [ApiDocsDefinition.aa, 'AA'],
  [ApiDocsDefinition.idmapping, 'ID Mapping'],
  [ApiDocsDefinition.async, 'Async Download'],
]);
