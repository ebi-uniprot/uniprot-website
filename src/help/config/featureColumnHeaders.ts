import getLabelAndTooltip from '../../shared/utils/getLabelAndTooltip';
import * as logging from '../../shared/utils/logging';

import { UniProtKBColumn } from '../../uniprotkb/types/columnTypes';

const featureColumnHeaders: Partial<
  Record<UniProtKBColumn, { tooltip: string; articleId: string }>
> = {
  [UniProtKBColumn.ftActSite]: {
    tooltip: 'Amino acid(s) directly involved in the activity of the enzyme',
    articleId: 'act_site',
  },

  [UniProtKBColumn.ftBinding]: {
    tooltip:
      'Binding site for any chemical group (co-enzyme, prosthetic group, etc.)',
    articleId: 'binding',
  },

  [UniProtKBColumn.ftCaBind]: {
    tooltip: 'The position(s) of calcium binding region(s) within the protein',
    articleId: 'ca_bind',
  },

  [UniProtKBColumn.ftCarbohyd]: {
    tooltip:
      'Position and description of a covalently attached glycan group(s)',
    articleId: 'carbohyd',
  },

  [UniProtKBColumn.ftChain]: {
    tooltip: 'Description of the mature polypeptide following processing',
    articleId: 'chain',
  },

  [UniProtKBColumn.ftCoiled]: {
    tooltip: 'Positions of region of coiled coil',
    articleId: 'coiled',
  },

  [UniProtKBColumn.ftCompbias]: {
    tooltip:
      'Region of compositional bias within the protein and description of the particular amino acids that are over-represented within this regions',
    articleId: 'compbias',
  },

  [UniProtKBColumn.ftConflict]: {
    tooltip: 'Description of sequence discrepancies of unknown origin',
    articleId: 'conflict',
  },

  [UniProtKBColumn.ftCrosslnk]: {
    tooltip: 'Residues participating in covalent linkage(s) between proteins',
    articleId: 'crosslink',
  },

  [UniProtKBColumn.ftDisulfid]: {
    tooltip: 'Cysteine residues participating in disulfide bonds',
    articleId: 'disulfid',
  },

  [UniProtKBColumn.ftDnaBind]: {
    tooltip:
      'Description of the position and type of each DNA-binding domain present within the protein',
    articleId: 'dna_bind',
  },

  [UniProtKBColumn.ftDomain]: {
    tooltip: 'Position and type of each modular protein domain',
    articleId: 'domain',
  },

  [UniProtKBColumn.ftHelix]: {
    tooltip: 'Positions of the experimentally determined helical region(s)',
    articleId: 'helix',
  },

  [UniProtKBColumn.ftInitMet]: {
    tooltip:
      'Indicates if the initiator methionine is cleaved in the mature protein',
    articleId: 'init_met',
  },

  [UniProtKBColumn.ftIntramem]: {
    tooltip: 'Extent of a region in a membrane, which does not cross it',
    articleId: 'intramem',
  },

  [UniProtKBColumn.ftLipid]: {
    tooltip: 'Position and description of a covalently attached lipid group(s)',
    articleId: 'lipid',
  },

  [UniProtKBColumn.ftMetal]: {
    tooltip: 'Binding site for a metal ion',
    articleId: 'metal',
  },

  [UniProtKBColumn.ftModRes]: {
    tooltip:
      'Position and description of a post-translational modification (PTM) excluding lipids, glycans and protein cross-links',
    articleId: 'mod_res',
  },

  [UniProtKBColumn.ftMotif]: {
    tooltip:
      'Short (up to 20 amino acids) sequence motif of biological interest',
    articleId: 'motif',
  },

  [UniProtKBColumn.ftMutagen]: {
    tooltip:
      'Position of the site which has been experimentally altered by mutagenesis and description of the associated phenotype',
    articleId: 'mutagen',
  },

  [UniProtKBColumn.ftNonCons]: {
    tooltip: 'Indicates that two residues in the sequence are not consecutive',
    articleId: 'non_cons',
  },

  [UniProtKBColumn.ftNpBind]: {
    tooltip: 'Region in the protein which binds nucleotide phosphates',
    articleId: 'np_bind',
  },

  [UniProtKBColumn.ftNonStd]: {
    tooltip:
      'Occurence of non-standard amino acid (selenocysteine or pyrrolysine) in the protein sequence',
    articleId: 'non_std',
  },

  [UniProtKBColumn.ftNonTer]: {
    tooltip:
      'Indicates that the residue is not terminal as the protein is not complete',
    articleId: 'non_ter',
  },

  [UniProtKBColumn.ftPeptide]: {
    tooltip: 'Position and length of the active peptide in the mature protein',
    articleId: 'peptide',
  },

  [UniProtKBColumn.ftPropep]: {
    tooltip:
      'Part of the protein that is cleaved during maturation or activation',
    articleId: 'propep',
  },

  [UniProtKBColumn.ftRegion]: {
    tooltip: 'Region of interest in the sequence',
    articleId: 'region',
  },

  [UniProtKBColumn.ftRepeat]: {
    tooltip:
      'Positions and types of repeated sequence motifs or repeated domains within the protein',
    articleId: 'repeat',
  },

  [UniProtKBColumn.ftSignal]: {
    tooltip:
      'Extent of the sequence targeting proteins to the secretory pathway or periplasmic space',
    articleId: 'signal',
  },

  [UniProtKBColumn.ftSite]: {
    tooltip: 'Interesting single amino acid site on the sequence',
    articleId: 'site',
  },

  [UniProtKBColumn.ftStrand]: {
    tooltip: 'Positions of the experimentally determined beta strand region(s)',
    articleId: 'strand',
  },

  [UniProtKBColumn.ftTopoDom]: {
    tooltip:
      'Description of the subcellular compartment where each non-membrane region of a membrane-spanning protein is found',
    articleId: 'topo_dom',
  },

  [UniProtKBColumn.ftTransit]: {
    tooltip: 'Extent of a transit peptide for organelle targeting',
    articleId: 'transit',
  },

  [UniProtKBColumn.ftTransmem]: {
    tooltip:
      'Description of the extent of a membrane-spanning region of the protein',
    articleId: 'transmem',
  },

  [UniProtKBColumn.ftTurn]: {
    tooltip:
      'Positions of the experimentally determined &apos;Turn&apos; region(s)',
    articleId: 'turn',
  },

  [UniProtKBColumn.ftUnsure]: {
    tooltip: 'Regions of uncertainty in the sequence',
    articleId: 'unsure',
  },

  [UniProtKBColumn.ftVarSeq]: {
    tooltip:
      'Amino acid change(s) due to alternative splicing (or other events) producing alternate protein isoforms',
    articleId: 'var_seq',
  },

  [UniProtKBColumn.ftVariant]: {
    tooltip: 'Description of a natural variant of the protein',
    articleId: 'variant',
  },

  [UniProtKBColumn.ftZnFing]: {
    tooltip: 'Position(s) and type(s) of zinc fingers within the protein',
    articleId: 'zn_fing',
  },
};

const getFeatureLabelAndTooltip = (
  label: string,
  featureColumn: UniProtKBColumn
) => {
  const header = featureColumnHeaders[featureColumn];
  if (header) {
    return getLabelAndTooltip(label, header.tooltip, header.articleId);
  }
  logging.warn(
    `No feature label information found for column: ${featureColumn}`
  );
  return { label };
};

export default getFeatureLabelAndTooltip;
