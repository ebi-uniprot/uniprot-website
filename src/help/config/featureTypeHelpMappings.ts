import FeatureType from '../../uniprotkb/types/featureType';

const FeatureTypeHelpMappings: Omit<Record<FeatureType, string>, 'Other'> = {
  'Compositional bias': 'compbias',
  'Non-standard residue': 'non_std',
  'Sequence uncertainty': 'unsure',
  'Sequence conflict': 'conflict',
  'Non-adjacent residues': 'non_cons',
  'Non-terminal residue': 'non_ter',
  'Alternative sequence': 'var_seq',
  Mutagenesis: 'mutagen',
  Domain: 'domain',
  Repeat: 'repeat',
  'Zinc finger': 'zn_fing',
  'DNA binding': 'dna_bind',
  Region: 'region',
  'Coiled coil': 'coiled',
  Motif: 'motif',
  'Active site': 'act_site',
  'Binding site': 'binding',
  Site: 'site',
  Intramembrane: 'intramem',
  'Topological domain': 'topo_dom',
  Transmembrane: 'transmem',
  'Initiator methionine': 'init_met',
  Signal: 'signal',
  'Transit peptide': 'transit',
  Propeptide: 'propep',
  Chain: 'chain',
  Peptide: 'peptide',
  'Modified residue': 'mod_res',
  Lipidation: 'lipid',
  Glycosylation: 'carbohyd',
  'Disulfide bond': 'disulfid',
  'Cross-link': 'crosslnk',
  Helix: 'helix',
  'Beta strand': 'strand',
  Turn: 'turn',
  'Natural variant': 'variant',
};

export default FeatureTypeHelpMappings;
