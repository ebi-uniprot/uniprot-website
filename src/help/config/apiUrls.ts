import joinUrl from 'url-join';

import { stringifyUrl } from '../../shared/utils/url';

import { apiPrefix } from '../../shared/config/apiUrls/apiPrefix';
import { defaultFacets as helpDefaultFacets } from './HelpFacetConfiguration';
import { ApiDocsDefinition } from '../types/apiDocumentation';

const entry = (id?: string) => id && joinUrl(apiPrefix, 'help', id);

const search = ({
  query,
  sort,
  fields,
  queryFacets,
  facets = helpDefaultFacets,
  size,
}: Record<string, string[] | string | null>) =>
  stringifyUrl(joinUrl(apiPrefix, 'help/search'), {
    query: [
      query || '*',
      ...(Array.isArray(queryFacets)
        ? queryFacets
        : (queryFacets || '').split(',')
      )
        .filter(Boolean)
        // Sort in order to improve cache hits
        .sort()
        .map((facet) => {
          const [facetName, facetValue] = (facet || '').split(':');
          return `(${facetName}:${
            facetValue.includes(' ') ? `"${facetValue}"` : facetValue
          })`;
        }),
    ]
      .filter(Boolean)
      .join(' AND '),
    sort,
    fields,
    facets,
    size,
  });

const apiDocumentationDefinition = (d: ApiDocsDefinition) =>
  joinUrl(apiPrefix, d, 'api/docs');

export default { entry, search, apiDocumentationDefinition };
