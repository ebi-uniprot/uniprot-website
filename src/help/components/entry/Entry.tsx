import cn from 'classnames';
import { Card, Loader } from 'franklin-sites';
import { MouseEventHandler, useCallback, useEffect, useMemo } from 'react';
import {
  generatePath,
  Navigate,
  useLocation,
  useNavigate,
  useParams,
} from 'react-router';
import {
  Attributes,
  defaults,
  IOptions,
  Tag,
  Transformer,
} from 'sanitize-html';

import { Location, LocationToPath } from '../../../app/config/urls';
import ErrorHandler from '../../../shared/components/error-pages/ErrorHandler';
import HTMLHead from '../../../shared/components/HTMLHead';
import { SingleColumnLayout } from '../../../shared/components/layouts/SingleColumnLayout';
import apiUrls from '../../../shared/config/apiUrls/apiUrls';
import useDataApiWithStale from '../../../shared/hooks/useDataApiWithStale';
import helper from '../../../shared/styles/helper.module.scss';
import {
  cleanTextDefaultOptions,
  getTransformTags,
  HeadingLevels,
} from '../../../shared/utils/cleanText';
import { sendGtagEventOutboundLinkClick } from '../../../shared/utils/gtagEvents';
import * as logging from '../../../shared/utils/logging';
import { parseMarkdown } from '../../../shared/utils/markdown';
import parseDate from '../../../shared/utils/parseDate';
import { stringifyQuery } from '../../../shared/utils/url';
import helpApiUrls from '../../config/apiUrls';
import { HelpEntryResponse } from '../../types/apiModel';
import RelatedArticles from './RelatedArticles';
import styles from './styles/entry.module.scss';

const internalRE = /^(https?:)?\/\/www.uniprot.org\//i;
const sameAppURL = new RegExp(window.location.origin + BASE_URL, 'i');
// NOTE: in production, internalRE and sameAppURL should be the same

const aTransformer: Transformer = (_: string, attribs: Attributes) => {
  const output: Tag = {
    tagName: 'a',
    attribs: {
      ...attribs,
      title: 'link',
    },
  };
  const href = attribs.href?.replace(internalRE, BASE_URL);
  if (href) {
    output.attribs.href = href;
    // if external link
    if (href === attribs.href && !href.startsWith('#')) {
      output.attribs.class = styles.external;
      output.attribs.target = '_blank';
      output.attribs.rel = 'noopener noreferrer';
    }
  }
  return output;
};

const allowedClasses = (cleanTextDefaultOptions.allowedClasses?.['*'] ||
  []) as string[];

// TODO: probably need to play with the options here in order to make it look OK
export const getCleanTextOptions = (headingLevel: HeadingLevels): IOptions => ({
  ...cleanTextDefaultOptions,
  allowedTags: [...defaults.allowedTags, 'img'],
  // none by default, so explicitely accept only the ones from the stylesheets
  allowedClasses: {
    '*': [
      ...allowedClasses,
      ...Object.values(helper),
      ...Object.values(styles),
    ],
  },
  allowedAttributes: {
    ...cleanTextDefaultOptions.allowedAttributes,
    img: ['src', 'width', 'height', 'alt'],
  },
  transformTags: {
    ...getTransformTags(headingLevel),
    a: aTransformer,
  },
});

type HelpEntryContentProps = {
  data: HelpEntryResponse;
  upperHeadingLevel?: HeadingLevels;
};

const HelpEntryContent = ({
  data,
  upperHeadingLevel = 'h1',
}: HelpEntryContentProps) => {
  const location = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    document.getElementById(location.hash.substring(1))?.scrollIntoView();
  }, [location.hash]);

  // Hijack clicks on content
  const handleClick = useCallback<MouseEventHandler<HTMLElement>>(
    (event) => {
      // Hijack clicks on content but only interested in link clicks
      if (event.target instanceof HTMLAnchorElement) {
        const { href } = event.target;
        // If clicks are within the UniProt client
        if (sameAppURL.test(href)) {
          if (event.metaKey || event.ctrlKey || event.shiftKey) {
            return; // default behaviour of opening a new tab or new window
          }
          // Don't navigate away!
          event.preventDefault();
          // And just replace the current URL with the next page
          navigate(href.replace(sameAppURL, '/'));
        } else {
          // analytics, similar as in InstrumentedExternalLink
          const url = new URL(href);
          sendGtagEventOutboundLinkClick(url.toString());
        }
      }
    },
    [navigate]
  );

  const html = useMemo(() => {
    if (data?.content) {
      return parseMarkdown(
        data.content,
        getCleanTextOptions(upperHeadingLevel)
      );
    }
    return null;
  }, [data, upperHeadingLevel]);

  if (!html) {
    return <ErrorHandler />;
  }

  // event delegation here, not actually doing anything with the div
  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
    <div dangerouslySetInnerHTML={{ __html: html }} onClick={handleClick} />
  );
};

type Props = {
  inPanel?: boolean;
  overrideContent?: HelpEntryResponse;
};

const HelpEntry = ({ inPanel, overrideContent }: Props) => {
  const { accession } = useParams();
  const navigate = useNavigate();

  const location = useLocation();
  const isReleaseNotes = location.pathname.includes('release-notes');

  let url = null;
  if (!overrideContent) {
    if (isReleaseNotes) {
      url = apiUrls.releaseNotes.entry(accession);
    } else {
      url = helpApiUrls.entry(accession);
    }
  }

  useEffect(() => {
    if (!isReleaseNotes || !accession || !accession.includes('/')) {
      return;
    }
    // if the accession has slashes, replace with dashes
    navigate(
      {
        ...location,
        pathname: generatePath(LocationToPath[Location.ReleaseNotesEntry], {
          accession: accession.replaceAll('/', '-'),
        }),
      },
      { replace: true }
    );
  }, [isReleaseNotes, accession, navigate, location]);

  const {
    data: loadedData,
    loading,
    error,
    progress,
    isStale,
  } = useDataApiWithStale<HelpEntryResponse>(url);

  useEffect(() => {
    if (inPanel && error) {
      logging.warn('contextual help article loading error', {
        extra: { data: accession },
      });
    }
  }, [inPanel, error, accession]);

  const data = overrideContent || loadedData;

  const date = useMemo(() => {
    if (data?.content) {
      return parseDate(isReleaseNotes ? data.releaseDate : data.lastModified);
    }
    return null;
  }, [data, isReleaseNotes]);

  if (loading && !data) {
    return <Loader progress={progress} />;
  }

  if (error || !data) {
    return (
      <Navigate
        replace
        to={{
          pathname: LocationToPath[Location.HelpResults],
          search: stringifyQuery({
            query: accession?.replaceAll('_', ' ') || '*',
          }),
        }}
      />
    );
  }

  if (inPanel) {
    return (
      <div className={styles['in-panel']}>
        <h2 className="medium">{data.title}</h2>
        <HelpEntryContent data={data} upperHeadingLevel="h3" />
      </div>
    );
  }

  const dateNode = date && (
    <div className={isReleaseNotes ? undefined : styles['last-updated-help']}>
      <small>
        {' '}
        {isReleaseNotes ? 'Released on' : 'Page last modified'}:{' '}
        <time dateTime={date.toISOString()}>{date.toDateString()}</time>
      </small>
    </div>
  );

  return (
    <SingleColumnLayout>
      <HTMLHead title={[data.title, 'UniProt help']} />
      <h1 className={data.categories?.includes('faq') ? 'big' : undefined}>
        {data.title}
      </h1>
      {isReleaseNotes && accession !== 'forthcoming-changes' && dateNode}
      <Card className={cn(styles.content, { [helper.stale]: isStale })}>
        <HelpEntryContent data={data} />
      </Card>
      {!isReleaseNotes && dateNode}
      {!loading && !isReleaseNotes && accession && data.categories?.length ? (
        <RelatedArticles current={accession} categories={data.categories} />
      ) : null}
    </SingleColumnLayout>
  );
};

export default HelpEntry;
