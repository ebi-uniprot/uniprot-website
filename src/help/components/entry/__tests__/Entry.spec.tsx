import { screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import customRender from '../../../../shared/__test-helpers__/customRender';

import Entry from '../Entry';

import { LocationToPath, Location } from '../../../../app/config/urls';

import mockData from '../../__mocks__/helpEntryModelData';

const mock = new MockAdapter(axios);

mock.onGet(/api\/help\/canonical_and_isoforms/).reply(200, mockData);

jest.mock('../../../../shared/components/layouts/UniProtFooter', () => ({
  __esModule: true,
  default: () => '{{ Footer }}',
}));

describe('Help entry tests', () => {
  it('should render the Entry component', async () => {
    const { asFragment } = customRender(
      <Route path={LocationToPath[Location.HelpEntry]} component={Entry} />,
      {
        route: '/help/canonical_and_isoforms',
      }
    );
    await screen.findByText(mockData.title);
    expect(asFragment()).toMatchSnapshot();
  });
});
