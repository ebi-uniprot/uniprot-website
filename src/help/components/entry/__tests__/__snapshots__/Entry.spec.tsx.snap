// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help entry tests should render the Entry component 1`] = `
<DocumentFragment>
  <div
    class="single-column-layout"
  >
    <main>
      <div
        class="uniprot-grid-cell--span-12 help-entry message message--warning"
        role="status"
      >
        <div
          class="message__side-border"
        />
        <test-file-stub
          classname="message--icon-align-top"
          height="1.125em"
          width="1.125em"
        />
        <div
          class="message__title"
        >
          Our Proteomes and UniProtKB/TrEMBL resources are undergoing a significant transition. Read our 
          <a
            href="/help/refprot_only_changes"
          >
            help page
          </a>
          , view 
          <a
            class="external-link no-right-margin"
            href="https://ftp.ebi.ac.uk/pub/contrib/UniProt/proteomes/"
            rel="noopener"
            target="_blank"
          >
            affected entries and proteomes
            <test-file-stub
              data-testid="external-link-icon"
              width="12.5"
            />
          </a>
          , or 
          <a
            href="/contact"
          >
            contact us
          </a>
           with any questions.
        </div>
      </div>
      <h1
        class="big"
      >
        What is the canonical sequence? Are all isoforms described in one entry?
      </h1>
      <section
        class="content card"
      >
        <div
          class="card__container"
        >
          <div
            class="card__content"
          >
            <div>
              <h2
                id="what-is-the-canonical-sequence"
              >
                What is the canonical sequence?
              </h2>
              

              <p>
                Each UniProtKB/Swiss-Prot entry contains all curated protein products encoded by a given gene in a given species or strain. For each UniProtKB/Swiss-Prot entry, we choose a canonical (or representative) sequence for display that should conform to at least one of the following criteria:
              </p>
              

              <ol>
                

                <li>
                  It is functional;
                </li>
                

                <li>
                  It is widely expressed;
                </li>
                

                <li>
                  It is encoded by conserved exons found in orthologous sequences;
                </li>
                

                <li>
                  It is identical to consensus sequences chosen by other resources and genome curation efforts such as CCDS and MANE (see also 
                  <a
                    href="/help/human_proteome"
                  >
                    UniProt's human proteome
                  </a>
                  )
                </li>
                

                <li>
                  In the absence of any information, we choose the longest sequence.
                </li>
                

              </ol>
              

              <p>
                Sequences chosen according to these criteria generally allow the description of the majority of functionally important 
                <a
                  href="/help/domain"
                >
                  domains
                </a>
                , 
                <a
                  href="/help/motif"
                >
                  motifs
                </a>
                , 
                <a
                  href="/help/site"
                >
                  sites
                </a>
                , and 
                <a
                  href="/help/ptm_processing_section"
                >
                  post-translational modifications
                </a>
                , naturally occurring 
                <a
                  href="/help/variant"
                >
                  variants
                </a>
                 with functional and clinical significance, and other 
                <a
                  href="/help/sequence_annotation"
                >
                  sequence features
                </a>
                .
              </p>
              

              <p>
                Additional information can be found in the 
                <a
                  href="/help/var_seq"
                >
                  'Alternative sequence'
                </a>
                 subsection.
              </p>
              

              <p>
                The various UniProtKB distribution formats (Flat Text, XML, RDF/XML) display only the canonical sequence. The website's 
                <a
                  href="/help/sequences_section"
                >
                  'Sequences'
                </a>
                 section displays the canonical sequence, but for convenience it offers also a view of the isoforms that are described in the 'Alternative sequence' subsection.
              </p>
              

              <h2
                id="are-all-isoforms-described-in-one-uniprotkbswiss-prot-entry"
              >
                Are all isoforms described in one UniProtKB/Swiss-Prot entry?
              </h2>
              

              <p>
                Whenever possible, all the protein products encoded by one gene in a given species are described in a single UniProtKB/Swiss-Prot entry, including isoforms generated by alternative splicing, alternative promoter usage, and alternative translation initiation (*). However, some alternative splicing isoforms derived from the same gene share only a few exons, if any at all, the same for some 'trans-splicing' events. In these cases, the divergence is obviously too important to merge all protein sequences into a single entry and the isoforms have to be described in separate 'external' entries.
              </p>
              

              <p>
                Example: 
                <a
                  href="/uniprotkb/P42284#sequences"
                >
                  isoforms derived from the lola gene (Drosophila melanogaster)
                </a>
              </p>
              

              <p>
                (*) Important remark: Due to the increase of sequence data coming from large-scale sequencing projects, UniProtKB/TrEMBL may contain additional predicted sequences encoded by genes which are described in a UniProtKB/Swiss-Prot entry.
              </p>
              

              <h2
                id="how-can-i-retrieve-isoform-sequences"
              >
                How can I retrieve isoform sequences?
              </h2>
              

              <p>
                Alternative sequences, described in either single or separate entries, are all available for Blast searches.
              </p>
              

              <p>
                Isoform sequences can be downloaded in FASTA format from our 
                <a
                  href="/downloads"
                >
                  FTP download index page
                </a>
                 (choose the file: 'Isoform sequences').
              </p>
              

              <p>
                Query-derived sets of canonical sequences alone or canonical and isoform sequences can also be downloaded in FASTA format (see 
                <a
                  href="/help/retrieve_sets"
                >
                  How to retrieve sets of UniProtKB protein sequences?
                </a>
                ).
              </p>
              

              <h2
                id="see-also"
              >
                See also
              </h2>
              

              <ul>
                

                <li>
                  <a
                    href="/help/cds_protein_definition"
                  >
                    What are UniProtKB's criteria for defining a CDS as a protein?
                  </a>
                </li>
                

                <li>
                  <a
                    href="/help/protein_diversity"
                  >
                    How are protein sequence variety and protein diversity represented in UniProtKB?
                  </a>
                </li>
                

                <li>
                  <a
                    href="/help/sequence_length"
                  >
                    Sequence length
                  </a>
                </li>
                

                <li>
                  <a
                    href="/help/alternative_products"
                  >
                    Alternative products
                  </a>
                </li>
                

                <li>
                  <a
                    href="/help/var_seq"
                  >
                    Alternative sequence
                  </a>
                </li>
                

                <li>
                  <a
                    href="/help/conflict"
                  >
                    Sequence conflict
                  </a>
                </li>
                

              </ul>
              

            </div>
          </div>
        </div>
      </section>
      <div
        class="last-updated-help"
      >
        <small>
           Page last modified: 
          <time
            datetime="2022-12-23T00:00:00.000Z"
          >
            Fri Dec 23 2022
          </time>
        </small>
      </div>
    </main>
  </div>
  {{ Footer }}
</DocumentFragment>
`;
